napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,521 @@
+    values: np.ndarray
+    current_value: Optional[Any] = None
+
+
+class ColorManager(EventedModel):
+    """A class for controlling the display colors for annotations in napari.
+
+    Attributes
+    ----------
+    current_color : Optional[np.ndarray]
+        A (4,) color array for the color of the next items to be added.
+    mode : ColorMode
+        The mode for setting colors.
+
+        ColorMode.DIRECT: colors are set by passing color values to ColorManager.colors
+        ColorMode.COLORMAP: colors are set via the continuous_colormap applied to the
+                            color_properties
+        ColorMode.CYCLE: colors are set vie the categorical_colormap appied to the
+                         color_properties. This should be used for categorical
+                         properties only.
+     color_properties : Optional[ColorProperties]
+        The property values that are used for setting colors in ColorMode.COLORMAP
+        and ColorMode.CYCLE. The ColorProperties dataclass has 3 fields: name,
+        values, and current_value. name (str) is the name of the property being used.
+        values (np.ndarray) is an array containing the property values.
+        current_value contains the value for the next item to be added. color_properties
+        can be set as either a ColorProperties object or a dictionary where the keys are
+        the field values and the values are the field values (i.e., a dictionary that would
+        be valid in ColorProperties(**input_dictionary) ).
+    continuous_colormap : Colormap
+        The napari colormap object used in ColorMode.COLORMAP mode. This can also be set
+        using the name of a known colormap as a string.
+    contrast_limits : Tuple[float, float]
+        The min and max value for the colormap being applied to the color_properties
+        in ColorMonde.COLORMAP mode. Set as a tuple (min, max).
+    categorical_colormap : CategoricalColormap
+        The napari CategoricalColormap object used in ColorMode.CYCLE mode.
+        To set a direct mapping between color_property values and colors,
+        pass a dictionary where the keys are the property values and the
+        values are colors (either string names or (4,) color arrays).
+        To use a color cycle, pass a list or array of colors. You can also
+        pass the CategoricalColormap keyword arguments as a dictionary.
+    colors : np.ndarray
+        The colors in a Nx4 color array, where N is the number of colors.
+    """
+
+    # fields
+    current_color: Optional[Array[float, (4,)]] = None
+    mode: ColorMode = ColorMode.DIRECT
+    color_properties: Optional[ColorProperties] = None
+    continuous_colormap: Colormap = 'viridis'
+    contrast_limits: Optional[Tuple[float, float]] = None
+    categorical_colormap: CategoricalColormap = [0, 0, 0, 1]
+    colors: Array[float, (-1, 4)] = []
+
+    # validators
+    @validator('continuous_colormap', pre=True)
+    def _ensure_continuous_colormap(cls, v):
```

Comment:
could this be a class validator on `Colormap` instead?

Reply:
Moving this on to `Colormap` causes a circular import because `ensure_colormap()` is defined in `colormap_utils.py`, which imports `Colormap`. Unfortunately, it's not as straightforward as moving `ensure_colormap()` to `colormap.py`, as there are other functions in `colormap_utils.py` that `ensure_colormap()` requires. I think we can either make `colormap.py` a mega file with all of the conversion stuff in it, refactor all of the conversion code in `colormap_utils.py` to not require `Colormap` (e.g., they return dicts of `Colormap` kwargs instead of `Colormap` objects`), or punt and consider a refactor later. I think punting is not so bad in this case, as the validator function is just calling `ensure_colormap`. It might be nice to get `ensure_colormap()` on `Colormap` in the future, as it will be a field in a few models around napari. What do you think, @tlambert03 and @sofroniewn ?

Reply:
I'd say just punt on it for now. 

I do think that ultimately that all of those functions in colormap_utils that just return `Colormap` instances would be nice to have in colormap itself as [classmethods](https://github.com/napari/napari/pull/1523#discussion_r471166439).  I could be convinced otherwise... but it might even be nice just to have this be part of the `Colormap.__new__` or `Colormap.__init__` process. (or, perhaps a few different `from_x` class methods that are dispatched depending on the type of the argument)